### ISSUE TITLE
Using Connection Pooling Queues with Python multiprocessing

### ISSUE THREAD

**Issue by palisadoes** [sqlalchemy/sqlalchemy#5460]
The SQLAlchemy documentation "[How do I use engines / connections / sessions with Python multiprocessing, or os.fork()?](https://docs.sqlalchemy.org/en/13/faq/connections.html#how-do-i-use-engines-connections-sessions-with-python-multiprocessing-or-os-fork)" does not consider using connection pooling queues with multiprocessing. 

The example given does not specify a Queue using a `poolclass` kwarg, therefore the default `None` queue is used. I have found when using queues that you need to **also** apply the `add_engine_pidguard` function to the imported queue class  **prior** to creating the engine or else the connections hang.

When using the stock example, some connections are logged as being checked out, but your application will hang eventually. This can be easily verified as you will see many process ids for your application in your process table when using the `ps aux | grep application_name` command that are not logged.

The resulting working code looks like this when using the `QueuePool` class.

**Note:** the add_engine_pidguard function is identical to the documentation:

```
import os
import warnings

from sqlalchemy import event
from sqlalchemy import exc
from sqlalchemy.pool import QueuePool

add_engine_pidguard(QueuePool)
engine = create_engine("...", poolclass=QueuePool)
add_engine_pidguard(engine)

def add_engine_pidguard(engine):
    """Add multiprocessing guards.

    Forces a connection to be reconnected if it is detected
    as having been shared to a sub-process.

    """

    @event.listens_for(engine, "connect")
    def connect(dbapi_connection, connection_record):
        connection_record.info['pid'] = os.getpid()

    @event.listens_for(engine, "checkout")
    def checkout(dbapi_connection, connection_record, connection_proxy):
        pid = os.getpid()
        if connection_record.info['pid'] != pid:
            # substitute log.debug() or similar here as desired
            warnings.warn(
                "Parent process %(orig)s forked (%(newproc)s) with an open "
                "database connection, "
                "which is being discarded and recreated." %
                {"newproc": pid, "orig": connection_record.info['pid']})
            connection_record.connection = connection_proxy.connection = None
            raise exc.DisconnectionError(
                "Connection record belongs to pid %s, "
                "attempting to check out in pid %s" %
                (connection_record.info['pid'], pid)
            )
```
----
**Versions (Multiple)**
 - OS: CentOS Linux release 7.8.2003 (Core)
 - Python: 3.6.8
 - SQLAlchemy: 1.3.15
 - Database: MariaDB-Galera-server-10.0.38-1
 - DBAPI: pymysql 0.9.3
----
 - OS: Ubuntu 20.04 LTS
 - Python: 3.8.2
 - SQLAlchemy: 1.3.16
 - Database: mariadb-server-10.3 (1:10.3.22-1ubuntu1)
 - DBAPI: pymysql 0.9.3
----

Please verify whether this observation is correct and whether my solution is officially recommended. I will update the documentation if this is so.

**Comment by zzzeek** [comment]
hi there -

I haven't looked at this recipe in awhile, but we use it [in production openstack](https://github.com/openstack/oslo.db/blob/master/oslo_db/sqlalchemy/engines.py#L387) and that covers an extremely wide array of usage by many organizations and very high scale, so a least within the way this recipe is deployed within Openstack, there are no known issues with hanging or anything else.     that's not to say there aren't deployment patterns that expose the issue you are seeing but would need very specific reproduction cases to understand that.

> The example given does not specify a Queue using a poolclass kwarg, therefore the default None queue is used.

So there is no "None" "Queue", the pool class for create_engine is QueuePool in all cases except when using SQLite, which uses NullPool where this issue doesn't really matter.   So first off if the recipe did not work for you until you specified QueuePool, that doesn't really add up, because it is already against a QueuePool for MySQL drivers.

> I have found when using queues that you need to also apply the add_engine_pidguard function to the imported queue class prior to creating the engine or else the connections hang.

So this has the effect of applying the event listeners to all occurrences of QueuePool within the process, so this hints that in your real world test there is perhaps another Engine/ QueuePool somewhere that is also being copied across the process boundary.     The engine.dispose() process also creates a new QueuePool however this transfers event listeners to the new pool.

> When using the stock example, some connections are logged as being checked out, but your application will hang eventually.

Can you please supply a fully self-contained, single file with no depenendencies other than Pymysql , working / reproducing example of this?   thanks.

**Comment by palisadoes** [comment]
The recipe works with the default `poolclass=NullPool`, it just doesn't work with `QueuePool`. I wasn't sufficiently precise, but it should be the equivalent of no pool being used correct?

Yes, the subprocesses use the Pool according to the guidelines in the [Session Basics, "When do I construct a Session, when do I commit it, and when do I close it?"](https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it) example, below:

```
from contextlib import contextmanager

@contextmanager
def session_scope():
    """Provide a transactional scope around a series of operations."""
    session = Session()
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()


def run_my_program():
    with session_scope() as session:
        ThingOne().go(session)
        ThingTwo().go(session)
```

Each subprocess can make multiple database queries during its lifetime.

I don't think standalone code will be necessary, as you have explained the most likely source of the issue. I think the documentation should at least identify this as an issue that could be encountered, especially when using another recommended pattern on the documentation website.

It took me a month of troubleshooting earlier in the year to figure out the solution, but didn't understand why it worked until now. I'd like to make a pull request to update the documentation in that "[How do I use engines / connections / sessions with Python multiprocessing, or os.fork()?](https://docs.sqlalchemy.org/en/13/faq/connections.html#how-do-i-use-engines-connections-sessions-with-python-multiprocessing-or-os-fork)" section to make it clearer. 

Let me know your thoughts.

**Comment by zzzeek** [comment]
> The recipe works with the default `poolclass=NullPool`, it just doesn't work with `QueuePool`.

as mentioned earlier, this recipe is in high volume production use with QueuePool for many years and no issues have been observed.

> I wasn't sufficiently precise, but it should be the equivalent of no pool being used correct?

The SQLAlchemy Engine always uses a "connection pool".  it just happens that there is one particular "pool" called "NullPool" that doesn't actually hold onto connections.    You can say that using "NullPool" is as if the Engine did not actually use a connection pool, yes.

> Each subprocess can make multiple database queries during its lifetime.

yes, however I am not seeing the significance of this statement.   Sessions use engines and they can make a lot of queries, even if they are in subprocesses, OK, and....that's what this recipe is intended for, and to my knowledge it does so without any known issues.

> 
> I don't think standalone code will be necessary, as you have explained the most likely source of the issue.
> I think the documentation should at least identify this as an issue that could be encountered, especially when using another recommended pattern on the documentation website.

Sorry, I'm not following.  No "issue" has been demonstrated or explained here.  


> 
> It took me a month of troubleshooting earlier in the year to figure out the solution, but didn't understand why it worked until now. I'd like to make a pull request to update the documentation in that "[How do I use engines / connections / sessions with Python multiprocessing, or os.fork()?](https://docs.sqlalchemy.org/en/13/faq/connections.html#how-do-i-use-engines-connections-sessions-with-python-multiprocessing-or-os-fork)" section to make it clearer.
> 
> Let me know your thoughts.

I think there might be a misunderstanding of subject matter here which is why [MCVE](https://stackoverflow.com/help/minimal-reproducible-example) are so important.     When you say, "it doesn't work with QueuePool", what does "doesn't work" mean?   you get a stack trace?  (if so please paste?)  you get a hang?  (if so please explain where the code is hung), etc.

**Comment by palisadoes** [comment]
Thanks. As mentioned before, the code hangs, specifically after spawning the sub-process. I will work on sample code as proof. Give me a day or two.

**Comment by zzzeek** [comment]
Please note I absolutely believe you are having a real issue and it's very likely the recipe is not accommodating for something, or otherwise something needs to be better documented.   it's just the QueuePool thing does not really tell me anything this was designed and tested for QueuePool.    thanks for your efforts.

**Comment by zzzeek** [comment]
so one thing I notice is that we have the same recipe twice and that's a problem, there are more docs here: https://docs.sqlalchemy.org/en/13/core/pooling.html#using-connection-pools-with-multiprocessing and I wrote the same recipe twice, which is weird, so let me move that.

next, here's how to make it hang.   you use up all the connections before the process starts.  that will make it hang for sure because the empty pool is transferred over.  the solution is to ensure engine.dispose() is called at the start of any subprocess.  the "pid" recipe is not really necessary if this is done correctly.


example:
```

from multiprocessing import Process
import os

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import exc

engine = create_engine(
    "mysql://scott:tiger@localhost/test",
    echo_pool="debug",
    max_overflow=0,
    pool_size=2,
)


@event.listens_for(engine, "connect")
def connect(dbapi_connection, connection_record):
    connection_record.info["pid"] = os.getpid()


@event.listens_for(engine, "checkout")
def checkout(dbapi_connection, connection_record, connection_proxy):
    pid = os.getpid()
    if connection_record.info["pid"] != pid:
        connection_record.connection = connection_proxy.connection = None
        raise exc.DisconnectionError(
            "Connection record belongs to pid %s, "
            "attempting to check out in pid %s"
            % (connection_record.info["pid"], pid)
        )


conn = engine.connect()
conn.close()


def run_in_process():
    print("Hi")
    with engine.connect() as conn:
        conn.execute("select 1")
    print("process is done")


p = Process(target=run_in_process)

c1 = engine.connect()
c2 = engine.connect()

# pool is empty

p.start()

p.join()

```

**Comment by palisadoes** [comment]
I've created a script the recreates the issue consistently. 

1. It uses a very scaled back version of our application using similar importations.
1. The script tries to do few DB updates in series before launching multiprocessing. This is similar to what the application does.
1. I've tried to use SQLAlchemy design patterns wherever possible.

Please follow these steps:

1. Put all the files below in the same directory.
1. Create a MySQL or MariaDB database with the `Language` table structure in `test.py`
1. Adjust your database authentication parameters in `__init__.py`
1. The log file is `/tmp/sqlalchemy-multiprocessing.log`. Tail this to see the pool disconnection messages, and some tracebacks.
1. `test.py` uses a number of useful flags
    1. `--verbose` will print the iteration count. I suggest you use this the first few times.
    1. `--iterations` will allow you to spawn X number of sub processes
    1. `--context` will use the python mulitprocessing `get_context` function. It works consistently when using this. The script works intermittently when this is not used.

I hope this will help to create new guidance.

--------
**test.py**
-------

```
#!/usr/bin/env python3
"""SQLAlchemy ORM tests script."""

# Main python libraries
import os
import random
from multiprocessing import get_context, Pool
import argparse

from sqlalchemy import UniqueConstraint, text
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.dialects.mysql import BIGINT, DATETIME
from sqlalchemy.dialects.mysql import VARBINARY
from sqlalchemy import Column
BASE = declarative_base()

#############################################################################
# Setup a global pool for database connections
#############################################################################
from __init__ import POOL
from db import db_modify


class Language(BASE):
    """Class defining the pt_language table of the database."""

    __tablename__ = 'pt_language'
    __table_args__ = (
        UniqueConstraint('code'),
        {'mysql_engine': 'InnoDB'}
    )

    idx_language = Column(
        BIGINT(unsigned=True), primary_key=True,
        autoincrement=True, nullable=False)

    code = Column(VARBINARY(512),
                  index=True, nullable=False, default=None)

    name = Column(
        VARBINARY(512),
        index=True, nullable=False, default=None)

    enabled = Column(
        BIGINT(unsigned=True), nullable=False, server_default='1')

    ts_modified = Column(
        DATETIME, server_default=text(
            'CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP'),)

    ts_created = Column(
        DATETIME, server_default=text('CURRENT_TIMESTAMP'))


def main():
    """Process agent data.

    Args:
        None

    Returns:
        None

    """
    # Process CLI
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--iterations', help='Number of database update iterations.',
        type=int,
        default=100)
    parser.add_argument(
        '--context', help='Use get_context multiprocessing function used.',
        action='store_true')
    parser.add_argument(
        '--verbose',
        help='''\
Log messages if used. Helps to prevent possibility of multiprocessing IO \
deadlocks when active.''',
        action='store_true')
    args = parser.parse_args()

    # Initialize key variables
    max_processes = 10
    arguments = [(x, args.verbose) for x in list(range(args.iterations))]

    for argument in arguments:
        run_in_process(argument, verbose=args.verbose)

    # Create a pool of sub process resources
    if args.context is True:
        with get_context('spawn').Pool(processes=max_processes) as pool:
            pool.starmap(run_in_process, arguments)
    else:
        with Pool(processes=max_processes) as pool:
            pool.starmap(run_in_process, arguments)

    # Wait for all the processes to end and get results
    pool.join()


def run_in_process(count, verbose=True):
    """Process agent data.

    Args:
        count: Process count
        verbose: Print messages if True

    Returns:
        None

    """
    # Initialize key variables
    enabled = random.getrandbits(1)

    # Process data
    pid = os.getpid()
    if bool(verbose):
        print('Starting process: {}, Count: {}'.format(pid, count))
    with db_modify() as session:
        session.query(Language).filter(
            Language.idx_language == 1).update(
                {'idx_language': 1,
                 'enabled': enabled})
    if bool(verbose):
        print('Stopping process: {}, Count: {}'.format(pid, count))


if __name__ == '__main__':
    # Run script
    main()

```
-------
**log.py**
-------
```
"""Module for logging."""

# Main python libraries
import logging

# Setup logging
LOG = logging.getLogger(__name__)
LOG_FILE = '/tmp/sqlalchemy-multiprocessing.log'
_LOG_HANDLER = logging.FileHandler(LOG_FILE)
_LOG_HANDLER.setFormatter(logging.Formatter('%(asctime)s %(message)s'))
_LOG_HANDLER.setLevel(logging.DEBUG)
LOG.addHandler(_LOG_HANDLER)
LOG.setLevel(logging.DEBUG)

```

--------
**db.py**
--------

```
"""Database interaction module."""

# Main python libraries
from contextlib import contextmanager
import sys
import traceback

# Test imports
from log import LOG, LOG_FILE
from __init__ import POOL


@contextmanager
def db_modify():
    """Provide a transactional scope around Update / Insert operations.

    From https://docs.sqlalchemy.org/en/13/orm/session_basics.html

    Args:
        error_code: Error code to use in messages
        die: Die if True

    Returns:
        None

    """
    # Initialize key variables
    prefix = 'Unable to modify database.'

    # Create session from pool
    session = POOL()

    # Setup basic functions
    try:
        yield session
        session.commit()
    except Exception as exception_error:
        session.rollback()
        log_message = '{}. Error: "{}"'.format(prefix, exception_error)
        LOG.debug(log_message)
    except:
        session.rollback()
        sys_exc_info = sys.exc_info()
        log_message = '{}. Unknown error'.format(prefix)
        db_logger(sys_exc_info, log_message)
    finally:
        # Return the Connection to the pool
        session.close()


@contextmanager
def db_query():
    """Provide a transactional scope around Query operations.

    From https://docs.sqlalchemy.org/en/13/orm/session_basics.html

    Args:
        error_code: Error code to use in messages

    Returns:
        None

    """
    # Initialize key variables
    prefix = 'Unable to read database.'

    # Create session from pool
    session = POOL()

    # Setup basic functions
    try:
        yield session
    except Exception as exception_error:
        session.close()
        log_message = '{}. Error: "{}"'.format(prefix, exception_error)
        LOG.debug(log_message)
    except:
        session.close()
        sys_exc_info = sys.exc_info()
        log_message = '{}. Unknown error'.format(prefix)
        db_logger(sys_exc_info, log_message)
    finally:
        # Return the Connection to the pool
        session.close()


def db_logger(sys_exc_info, message):
    """Log exception information.

    Args:
        sys_exc_info: Result of sys.exec_info()
        message: Message to log

    Returns:
        None

    """
    (exc_type, exc_value, exc_traceback) = sys_exc_info
    trace_info = ('''\
Bug: Exception Type:{}, Exception Instance: {}, Stack Trace Object: {}]\
'''.format(exc_type, exc_value, exc_traceback))
    LOG.debug(trace_info)
    LOG.debug(message)
    with open(LOG_FILE, 'a+') as _fh:
        traceback.print_tb(exc_traceback, file=_fh)

```
--------
**__init__.py**
--------

```
"""Module to setup SQLAlchemy connection pooling."""

# Main python libraries
import os
from contextlib import contextmanager
from multiprocessing import get_context
import logging
import sys

# PIP imports
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker, scoped_session
from sqlalchemy import event
from sqlalchemy import exc
from sqlalchemy.pool import QueuePool

# Setup logging
from log import LOG

#############################################################################
# Setup a global pool for database connections
#############################################################################
POOL = None


def main():
    """Process agent data.

    Args:
        None

    Returns:
        None

    """
    # Initialize key variables
    global POOL

    # Define SQLAlchemy parameters from configuration
    pool_timeout = 30
    pool_recycle = min(10, pool_timeout - 10)
    pool_size = 2
    max_overflow = 0

    db_url = '''\
mysql+pymysql://USERNAME:PASSWORD@localhost/DATABASE?charset=utf8mb4'''

    # Fix for multiprocessing on pools.
    _add_engine_pidguard(QueuePool)

    # Add MySQL to the pool
    db_engine = create_engine(
        db_url,
        echo=False,
        echo_pool=False,
        encoding='utf8',
        poolclass=QueuePool,
        max_overflow=max_overflow,
        pool_size=pool_size,
        pool_pre_ping=True,
        pool_recycle=pool_recycle,
        pool_timeout=pool_timeout)

    # Fix for multiprocessing on engines.
    _add_engine_pidguard(db_engine)

    # Create database session object
    POOL = scoped_session(
        sessionmaker(
            autoflush=True,
            autocommit=False,
            bind=db_engine
        )
    )


def _add_engine_pidguard(engine):
    """Add multiprocessing guards.

    Forces a connection to be reconnected if it is detected
    as having been shared to a sub-process.

    source
    ------

    http://docs.sqlalchemy.org/en/latest/faq/connections.html
    "How do I use engines / connections / sessions with
    Python multiprocessing, or os.fork()?"

    Args:
        engine: SQLalchemy engine instance

    Returns:
        None

    """
    @event.listens_for(engine, 'connect')
    def connect(dbapi_connection, connection_record):
        """Get the PID of the sub-process for connections.

        Args:
            dbapi_connection: A SqlALchemy DBAPI connection.
            connection_record: The SqlALchemy _ConnectionRecord managing the
                DBAPI connection.

        Returns:
            None

        """
        # Update the connection_record variable for later
        connection_record.info['pid'] = os.getpid()

    @event.listens_for(engine, 'checkout')
    def checkout(dbapi_connection, connection_record, connection_proxy):
        """Checkout sub-processes connection for sub-processing if needed.

            Checkout is called when a connection is retrieved from the Pool.

        Args:
            dbapi_connection: A SqlALchemy DBAPI connection.
            connection_record: The SqlALchemy _ConnectionRecord managing the
                DBAPI connection.
            connection_proxy: The SqlALchemy _ConnectionFairy object which
                will proxy the public interface of the DBAPI connection for the
                lifespan of the checkout.

        Returns:
            None

        """
        # Get PID of main process
        pid = os.getpid()

        # Detect if this is a sub-process
        if connection_record.info['pid'] != pid:
            log_message = ('''\
Parent process {} forked ({}) with an open database connection, \
which is being discarded and recreated.\
'''.format(connection_record.info['pid'], pid))
            LOG.debug(log_message)

            connection_record.connection = connection_proxy.connection = None
            raise exc.DisconnectionError('''\
Connection record belongs to pid {}, attempting to check out in pid {}\
'''.format(connection_record.info['pid'], pid))


if __name__ == '__init__':
    main()

```

**Comment by zzzeek** [comment]
hi there -

any chance this can be produced as an [MCVE](https://stackoverflow.com/help/minimal-reproducible-example)?  this is many files and seems to be an entire application.       in my [example](https://github.com/sqlalchemy/sqlalchemy/issues/5460#issuecomment-659059570) above I illustrate how to cause a deadlock if the engine is passed to a subprocess without calling dispose() first to ensure the pool is at a starting state.    I notice there is no call to .dispose() in your example.    This should be done on any engine as soon as the subprocess starts, and I have [updated](https://docs.sqlalchemy.org/en/13/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork) the documentation to make this clear.

if this solves your issue that would save a lot of time as I dont' really have the resources to debug full applications.

**Comment by zzzeek** [comment]
also it looks like you are reusing Session objects across requests.    don't use the scoped_session() here as this is potentially part of the problem.     reassign "POOL" to the sessionmaker directly so that when you call session.close(), that Session is gone.

**Comment by palisadoes** [comment]
POOL doesn't have dispose(). Where would you put it in this case?

**Comment by zzzeek** [comment]
on the engine, see https://docs.sqlalchemy.org/en/13/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork

**Comment by zzzeek** [comment]
is "get_context('spawn')" important?

**Comment by zzzeek** [comment]
is it necessary to run "run_in_process()" a bunch of times first and then do the multiprocessing thing?  im trying to cut this down to one file without anything extraneous.

**Comment by zzzeek** [comment]
this seems to need data in the table to work, not sure.  will try with a blank table

**Comment by palisadoes** [comment]
> is it necessary to run "run_in_process()" a bunch of times first and then do the multiprocessing thing? im trying to cut this down to one file without anything extraneous.

I did it to emulate possible database transactions before.

**Comment by palisadoes** [comment]
> is "get_context('spawn')" important?

Yes, the script runs without issue when you use it. You'll notice that the `pid_guard` isn't invoked when it's applied.

**Comment by zzzeek** [comment]
I have a shorter vrsion that does hang but if I do an engine.dispose() before entering the multiprocessing, it's fine.   im not sure yet the exact nature of the hang, it's clear that the pool has no connections available due to state carried over from teh parent is the cause but let me see if i can identify exactly.

**Comment by palisadoes** [comment]
[The updated documentation](https://docs.sqlalchemy.org/en/13/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork) should cover the use of connection pooling, because you have to use the dispose() on the engine, but you take use sessions from the pool, not the engine.

**Comment by zzzeek** [comment]
Hi - 

there is no ORM session in a "pool".   The ORM Session refers to an Engine which refers to a Pool.      make sure the engine is disposed before you call the child processes, and that solves the problem.  the issue seems like it may be PyMySQL having to clean up these connections within a weakref finalizer on the SQLAlchemy side and the "ROLLBACK" hangs.    If I use the native client, the problem also goes away.

this program will hang if you don't do the dispose(), and if you restore it, then it succeeds.  no event is needed.

```


from contextlib import contextmanager
from multiprocessing import Pool
import os
import random

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import text
from sqlalchemy import UniqueConstraint
from sqlalchemy.dialects.mysql import BIGINT
from sqlalchemy.dialects.mysql import DATETIME
from sqlalchemy.dialects.mysql import VARBINARY
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
BASE = declarative_base()


class Language(BASE):
    """Class defining the pt_language table of the database."""

    __tablename__ = 'pt_language'
    __table_args__ = (
        UniqueConstraint('code'),
        {'mysql_engine': 'InnoDB'}
    )

    idx_language = Column(
        BIGINT(unsigned=True), primary_key=True,
        autoincrement=True, nullable=False)

    code = Column(VARBINARY(512),
                  index=True, nullable=False, default=None)

    name = Column(
        VARBINARY(512),
        index=True, nullable=False, default=None)

    enabled = Column(
        BIGINT(unsigned=True), nullable=False, server_default='1')

    ts_modified = Column(
        DATETIME, server_default=text(
            'CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP'),)

    ts_created = Column(
        DATETIME, server_default=text('CURRENT_TIMESTAMP'))


def run_in_process(count, verbose=True):
    """Process agent data.

    Args:
        count: Process count
        verbose: Print messages if True

    Returns:
        None

    """
    # Initialize key variables
    enabled = random.getrandbits(1)

    # Process data
    pid = os.getpid()
    if bool(verbose):
        print('Starting process: {}, Count: {}'.format(pid, count))
    with db_modify() as session:
        session.query(Language).filter(
            Language.idx_language == 1).update(
                {'idx_language': 1,
                 'enabled': enabled})
    if bool(verbose):
        print('Stopping process: {}, Count: {}'.format(pid, count))


@contextmanager
def db_modify():

    # Create session from pool
    session = POOL()

    try:
        yield session
        session.commit()
    except Exception as exception_error:
        session.rollback()
        print(exception_error)
    except:
        session.rollback()
        raise
    finally:
        # Return the Connection to the pool
        session.close()

POOL = None


def main():
    """Process agent data.

    Args:
        None

    Returns:
        None

    """
    # Initialize key variables
    global POOL

    # Define SQLAlchemy parameters from configuration
    pool_timeout = 30
    pool_recycle = min(10, pool_timeout - 10)
    pool_size = 2
    max_overflow = 0

    db_url = '''\
mysql+pymysql://scott:tiger@localhost/test'''


    # Add MySQL to the pool
    db_engine = create_engine(
        db_url,
        echo=True, #False,
        echo_pool=True,
        encoding='utf8',
        max_overflow=max_overflow,
        pool_size=pool_size,
        pool_pre_ping=True,
        pool_recycle=pool_recycle,
        pool_timeout=pool_timeout)

    Language.metadata.drop_all(db_engine)
    Language.metadata.create_all(db_engine)

    # ensure connections are disposed before sharing engine.
    db_engine.dispose()

    POOL = sessionmaker(
        autoflush=True,
        autocommit=False,
        bind=db_engine
    )

    # Initialize key variables
    max_processes = 10
    arguments = [(x, False) for x in list(range(100))]

    with Pool(processes=max_processes) as pool:
        pool.starmap(run_in_process, arguments)

    # Wait for all the processes to end and get results
    pool.join()


if __name__ == '__main__':
    main()


```

**Comment by palisadoes** [comment]
This has been a learning experience for me, thanks for taking the time.

Why is there a difference in the location of dispose() in this link [Using Connection Pools with Multiprocessing or os.fork()](https://docs.sqlalchemy.org/en/13/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork) and your example above?

I favor having the dispose before invoking sessonmaker, as you don't have to remember to put dispose() as the first line of code in every multi-processed function. It helps to make the code more maintainable if others need to use the engine using the db_modify() function.

It would be good to explain the different strategies in your updated documentation.

**Comment by zzzeek** [comment]
the idea is that when an Engine/Pool is used in a child fork, it should have no database connections present from the parent process when it is used.

One way to do this is to call engine.dispose() **before** the child process is forked.   That way when the fork occurs, the Engine / Pool in the child process will have no database connections.  This is the approach taken in how I altered your example, which I used because I felt it was simple to understand.

The other way is to call engine.dispose() **after** the child process is forked, but **before** the engine/pool is used in that child process.  This is the approach taken in the docs, which is used because it makes no assumptions about how the child process came about and is therefore more likely to be successful.

in my opinion, which approach is used is not as important for this documentation as much as it is that the concept of what is going on is explained, which is that an engine should have its pooled connections disposed before it is used in a child process.
